moby/moby
# Add support for non-swarm container secrets

Closes #40046.

- What I did
	- [x] Add support for non-swarm container secrets using APIs
	- [ ] Add support for non-swarm container secrets using command line

- How I did it
	- Added `Secrets` field to the `github.com/docker/docker/api/types/container.Config` type
	- Include secret file creation on host machine, and secrets mounting procedures to container's start time (similar to how Swarm secrets are created and mounted to the containers)

- How to verify it
	- Direct Daemon API Interaction
		- Call `ContainerCreate` daemon API with the request body containing the newly added `Config.Secrets` field
		- Check `Config.Secrets` section of the container's `config.v2.json` file to see secrets associated to the container
		- Starting the container will
			- Create secret files in `/mounts/secrets/<secret-id>` of the container root directory, similar to Swarm secrets
			- Mount secret files to defined paths in the container
	- Using Docker-Compose
		- Download and build forked `docker-compose` project from [here](https://github.com/xeptore/compose)
		- Use the sample compose configuration and example `password.txt` secret file below, you can verify that the secrets defined for the service, are being mounted to the container with proper permission and mode:
			- `compose.yml`:
				```yml
				version: "3.9"
				services:
				  web:
				    image: nginx:latest
				    secrets:
				      - source: password
				        target: /var/run/my_secret
				        mode: 0400
				        uid: "95"
				        gid: "85"
				      - password
				secrets:
				  password:
				    file: ./password.txt
				    name: password
				```
			- `password.txt`
				```txt
				p4sSw0rD
				```

- Description for the changelog
	- The proposed solution is similar to how secrets are being treated in Swarm mode.
	- I decided to store secret raw file content in the container's `config.v2.json` file, instead of storing them in an in-memory or file-based database, without any encryption/decryption mechanism, as it's a local container, and file permissions seems enough for me. Also, it's simpler!
	- Current implementation should be backward-compatible as it only adds a new field to the container config field, which if not used, does not change anything for the container, even contents of `config.v2.json` file.
	- Current implementation does not affect Swarm mode secrets, as the fields that hold information about secrets are different.
	- I tried to reduce code duplication between Swarm mode and non-Swarm mode secret management, but if there's anything that can be improved, I'm willing to refactor the changes.

- Considerations
	- Other than the security vulnerabilities that storing container secrets in its `config.v2.json` file might introduce, I cannot see any other major issues
